File: src/renderer/vulkan/MeshCache.cpp
////////////////////////////////////////////////////////////////////////////////
/// Reaper
///
/// Copyright (c) 2015-2022 Thibault Schueller
/// This file is distributed under the MIT License
////////////////////////////////////////////////////////////////////////////////

#include "MeshCache.h"

#include <vulkan_loader/Vulkan.h>

#include "renderer/vulkan/Backend.h"

#include "mesh/Mesh.h"

#include <meshoptimizer.h>

#include "renderer/shader/meshlet/meshlet.share.hlsl"

namespace Reaper
{
MeshCache create_mesh_cache(VulkanBackend& backend)
{
    MeshCache cache;

    cache.indexBuffer = create_buffer(
        backend.device, "Index buffer",
        DefaultGPUBufferProperties(MeshCache::MAX_INDEX_COUNT, sizeof(int), GPUBufferUsage::StorageBuffer),
        backend.vma_instance, MemUsage::CPU_To_GPU);

    cache.vertexBufferPosition = create_buffer(
        backend.device, "Position buffer",
        DefaultGPUBufferProperties(MeshCache::MAX_VERTEX_COUNT, 3 * sizeof(float), GPUBufferUsage::StorageBuffer),
        backend.vma_instance, MemUsage::CPU_To_GPU);

    cache.vertexAttributesBuffer =
        create_buffer(backend.device, "Vertex attributes",
                      DefaultGPUBufferProperties(MeshCache::MAX_VERTEX_COUNT, sizeof(VertexAttributes),
                                                 GPUBufferUsage::StorageBuffer),
                      backend.vma_instance, MemUsage::CPU_To_GPU);

    cache.meshletBuffer = create_buffer(
        backend.device, "Meshlet buffer",
        DefaultGPUBufferProperties(MeshCache::MAX_MESHLET_COUNT, sizeof(Meshlet), GPUBufferUsage::StorageBuffer),
        backend.vma_instance, MemUsage::CPU_To_GPU);

    clear_meshes(cache);

    return cache;
}

void destroy_mesh_cache(VulkanBackend& backend, const MeshCache& mesh_cache)
{
    vmaDestroyBuffer(backend.vma_instance, mesh_cache.indexBuffer.handle, mesh_cache.indexBuffer.allocation);
    vmaDestroyBuffer(backend.vma_instance, mesh_cache.vertexBufferPosition.handle,
                     mesh_cache.vertexBufferPosition.allocation);
    vmaDestroyBuffer(backend.vma_instance, mesh_cache.vertexAttributesBuffer.handle,
                     mesh_cache.vertexAttributesBuffer.allocation);
    vmaDestroyBuffer(backend.vma_instance, mesh_cache.meshletBuffer.handle, mesh_cache.meshletBuffer.allocation);
}

void clear_meshes(MeshCache& mesh_cache)
{
    mesh_cache.mesh2_instances.clear();

    mesh_cache.current_index_offset = 0;
    mesh_cache.current_position_offset = 0;
    mesh_cache.current_attributes_offset = 0;
    mesh_cache.current_meshlet_offset = 0;
}

namespace
{
    MeshAlloc mesh_cache_allocate_mesh(MeshCache& mesh_cache, const Mesh& mesh, std::span<const Meshlet> meshlets)
    {
        MeshAlloc alloc = {};

        const u32 index_count = static_cast<u32>(mesh.indexes.size());
        const u32 position_count = static_cast<u32>(mesh.positions.size());
        const u32 attributes_count = static_cast<u32>(mesh.attributes.size());
        const u32 meshlet_count = static_cast<u32>(meshlets.size());

        Assert(index_count > 0);
        Assert(position_count > 0);
        Assert(position_count == attributes_count);
        Assert(meshlet_count > 0);

        alloc.index_count = index_count;
        alloc.vertex_count = position_count;
        alloc.meshlet_count = meshlet_count;

        alloc.index_offset = mesh_cache.current_index_offset;
        alloc.position_offset = mesh_cache.current_position_offset;
        alloc.attributes_offset = mesh_cache.current_attributes_offset;
        alloc.meshlet_offset = mesh_cache.current_meshlet_offset;

        mesh_cache.current_index_offset += index_count;
        mesh_cache.current_position_offset += position_count;
        mesh_cache.current_attributes_offset += position_count;
        mesh_cache.current_meshlet_offset += meshlet_count;

        Assert(mesh_cache.current_index_offset < MeshCache::MAX_INDEX_COUNT);
        Assert(mesh_cache.current_position_offset < MeshCache::MAX_VERTEX_COUNT);
        Assert(mesh_cache.current_attributes_offset < MeshCache::MAX_VERTEX_COUNT);
        Assert(mesh_cache.current_meshlet_offset < MeshCache::MAX_MESHLET_COUNT);

        return alloc;
    }

    void upload_mesh_to_mesh_cache(MeshCache& mesh_cache, const Mesh& mesh, const MeshAlloc& mesh_alloc,
                                   std::span<const Meshlet> meshlets, VulkanBackend& backend)
    {
        upload_buffer_data_deprecated(backend.device, backend.vma_instance, mesh_cache.indexBuffer, mesh.indexes.data(),
                                      mesh.indexes.size() * sizeof(mesh.indexes[0]), mesh_alloc.index_offset);
        upload_buffer_data_deprecated(backend.device, backend.vma_instance, mesh_cache.vertexBufferPosition,
                                      mesh.positions.data(), mesh.positions.size() * sizeof(mesh.positions[0]),
                                      mesh_alloc.position_offset);
        upload_buffer_data_deprecated(backend.device, backend.vma_instance, mesh_cache.vertexAttributesBuffer,
                                      mesh.attributes.data(), mesh.attributes.size() * sizeof(mesh.attributes[0]),
                                      mesh_alloc.attributes_offset);

        upload_buffer_data_deprecated(backend.device, backend.vma_instance, mesh_cache.meshletBuffer, meshlets.data(),
                                      meshlets.size() * sizeof(meshlets[0]), mesh_alloc.meshlet_offset);
    }
} // namespace

void load_meshes(VulkanBackend& backend, MeshCache& mesh_cache, std::span<const Mesh> meshes,
                 std::span<MeshHandle> output_handles)
{
    Assert(output_handles.size() >= meshes.size());

    for (u32 mesh_index = 0; mesh_index < meshes.size(); mesh_index++)
    {
        // NOTE: Deep copy to keep the input immutable
        Mesh mesh = meshes[mesh_index];

        Assert(mesh.attributes.size() == mesh.positions.size());

        const size_t max_vertices = MeshletMaxTriangleCount * 3;
        const size_t max_triangles = MeshletMaxTriangleCount;
        const float  cone_weight = 0.5f; // FIXME fold in a constant

        Assert(max_vertices < 256); // NOTE: u8 indices

        size_t max_meshlets = meshopt_buildMeshletsBound(mesh.indexes.size(), max_vertices, max_triangles);

        std::vector<meshopt_Meshlet> meshlets(max_meshlets);
        std::vector<u32>             meshlet_vertices(max_meshlets * max_vertices);
        std::vector<u8>              meshlet_indices(max_meshlets * max_triangles * 3);

        size_t meshlet_count =
            meshopt_buildMeshlets(meshlets.data(), meshlet_vertices.data(), meshlet_indices.data(), mesh.indexes.data(),
                                  mesh.indexes.size(), &mesh.positions[0].x, mesh.positions.size(),
                                  sizeof(mesh.positions[0]), max_vertices, max_triangles, cone_weight);

        const meshopt_Meshlet& last = meshlets[meshlet_count - 1];

        meshlet_vertices.resize(last.vertex_offset + last.vertex_count);
        meshlet_indices.resize(last.triangle_offset + last.triangle_count * 3);
        meshlets.resize(meshlet_count);

        const u32 meshlet_vertex_count = static_cast<u32>(meshlet_vertices.size());
        const u32 total_mesh_index_count = static_cast<u32>(meshlet_indices.size());

        std::vector<u32>              optimized_index_buffer(total_mesh_index_count);
        std::vector<glm::fvec3>       optimized_position_buffer(meshlet_vertex_count);
        std::vector<VertexAttributes> optimized_attributes_buffer(meshlet_vertex_count);
        std::vector<Meshlet>          optimized_meshlets(meshlet_count);

        for (u32 i = 0; i < meshlet_vertex_count; i++)
        {
            const u32 index = meshlet_vertices[i];

            optimized_position_buffer[i] = mesh.positions[index];
            optimized_attributes_buffer[i] = mesh.attributes[index];
        }

        u32 index_output_offset = 0;

        // We also do index buffer compaction in the same pass
        for (u32 meshlet_index = 0; meshlet_index < meshlets.size(); meshlet_index++)
        {
            const meshopt_Meshlet& meshlet = meshlets[meshlet_index];
            const meshopt_Bounds boundsMeshlet = meshopt_computeMeshletBounds(&meshlet_vertices[meshlet.vertex_offset],
                                                                              &meshlet_indices[meshlet.triangle_offset],
                                                                              meshlet.triangle_count,
                                                                              &mesh.positions[0].x,
                                                                              mesh.positions.size(),
                                                                              sizeof(mesh.positions[0]));

            const u32 meshlet_index_count = meshlet.triangle_count * 3;

            Meshlet& meshlet_instance = optimized_meshlets[meshlet_index];
            meshlet_instance.vertex_offset = meshlet.vertex_offset;
            meshlet_instance.vertex_count = meshlet.vertex_count;
            meshlet_instance.index_offset = index_output_offset;
            meshlet_instance.index_count = meshlet_index_count;
            meshlet_instance.center_ms =
                glm::fvec3(boundsMeshlet.center[0], boundsMeshlet.center[1], boundsMeshlet.center[2]);
            meshlet_instance.radius = boundsMeshlet.radius;
            meshlet_instance.cone_axis_ms =
                glm::fvec3(boundsMeshlet.cone_axis[0], boundsMeshlet.cone_axis[1], boundsMeshlet.cone_axis[2]);
            meshlet_instance.cone_cutoff = boundsMeshlet.cone_cutoff;
            meshlet_instance.cone_apex_ms =
                glm::fvec3(boundsMeshlet.cone_apex[0], boundsMeshlet.cone_apex[1], boundsMeshlet.cone_apex[2]);

            // Copy index buffer
            for (u32 index = 0; index < meshlet_index_count; index++)
            {
                const u32 index_input_offset = meshlet.triangle_offset + index;

                optimized_index_buffer[index_output_offset + index] =
                    static_cast<u32>(meshlet_indices[index_input_offset]);
            }

            index_output_offset += meshlet_index_count;
        }

        optimized_index_buffer.resize(index_output_offset); // Trim excess

        std::swap(mesh.indexes, optimized_index_buffer);
        std::swap(mesh.positions, optimized_position_buffer);
        std::swap(mesh.attributes, optimized_attributes_buffer);

        const MeshHandle new_handle = static_cast<MeshHandle>(mesh_cache.mesh2_instances.size());
        Mesh2&           mesh2 = mesh_cache.mesh2_instances.emplace_back();

        mesh2 = create_mesh2(mesh_cache_allocate_mesh(mesh_cache, mesh, optimized_meshlets));

        upload_mesh_to_mesh_cache(mesh_cache, mesh, mesh2.lods_allocs[0], optimized_meshlets, backend);

        output_handles[mesh_index] = new_handle;
    }
}
} // namespace Reaper

File: src/renderer/vulkan/mesh_cache.zig
// Port of src/renderer/vulkan/MeshCache.{h,cpp}
//
// One big buffer per stream (indices, positions, attributes, meshlets), all
// host-visible, with a bump allocator handing out sub-ranges. Loading a mesh
// runs it through meshoptimizer's clusterizer, reorders the vertex streams into
// meshlet order, and compacts the per-meshlet index runs into one buffer.

const std = @import("std");
const vk = @import("vulkan");

const buffer_module = @import("buffer.zig");
const gpu_buffer = @import("../buffer/gpu_buffer.zig");
const hlsl_meshlet = @import("../hlsl/meshlet/meshlet.zig");
const mesh_module = @import("../../mesh/mesh.zig");
const mesh2 = @import("../mesh2.zig");
const meshlet_builder = @import("../../mesh/meshlet_builder.zig");
const vma = @import("vma.zig").c;

pub const MeshletMesh = meshlet_builder.MeshletMesh;
pub const buildMeshlets = meshlet_builder.buildMeshlets;

const GPUBuffer = buffer_module.GPUBuffer;
const Mesh = mesh_module.Mesh;
const MeshAlloc = mesh2.MeshAlloc;
const MeshHandle = mesh2.MeshHandle;
const Meshlet = hlsl_meshlet.Meshlet;
const VertexAttributes = mesh_module.VertexAttributes;

const log = std.log.scoped(.renderer);

pub const max_index_count: u32 = 2_000_000;
pub const max_vertex_count: u32 = 4_000_000;
pub const max_meshlet_count: u32 = max_vertex_count / 64; // FIXME

pub const MeshCache = struct {
    index_buffer: GPUBuffer,
    vertex_buffer_position: GPUBuffer,
    vertex_attributes_buffer: GPUBuffer,
    meshlet_buffer: GPUBuffer,

    current_index_offset: u32 = 0,
    current_position_offset: u32 = 0,
    current_attributes_offset: u32 = 0,
    current_meshlet_offset: u32 = 0,

    mesh2_instances: std.ArrayList(mesh2.Mesh2) = .empty,

    allocator: std.mem.Allocator,

    pub fn init(vma_instance: vma.VmaAllocator, allocator: std.mem.Allocator) !MeshCache {
        const usage = gpu_buffer.GPUBufferUsage{ .storage_buffer = true };

        const index_buffer = try buffer_module.createBuffer(
            vma_instance,
            gpu_buffer.defaultBufferProperties(max_index_count, @sizeOf(u32), usage),
            .cpu_to_gpu,
        );
        errdefer buffer_module.destroyBuffer(vma_instance, index_buffer);

        const vertex_buffer_position = try buffer_module.createBuffer(
            vma_instance,
            gpu_buffer.defaultBufferProperties(max_vertex_count, 3 * @sizeOf(f32), usage),
            .cpu_to_gpu,
        );
        errdefer buffer_module.destroyBuffer(vma_instance, vertex_buffer_position);

        const vertex_attributes_buffer = try buffer_module.createBuffer(
            vma_instance,
            gpu_buffer.defaultBufferProperties(max_vertex_count, @sizeOf(VertexAttributes), usage),
            .cpu_to_gpu,
        );
        errdefer buffer_module.destroyBuffer(vma_instance, vertex_attributes_buffer);

        const meshlet_buffer = try buffer_module.createBuffer(
            vma_instance,
            gpu_buffer.defaultBufferProperties(max_meshlet_count, @sizeOf(Meshlet), usage),
            .cpu_to_gpu,
        );

        return .{
            .index_buffer = index_buffer,
            .vertex_buffer_position = vertex_buffer_position,
            .vertex_attributes_buffer = vertex_attributes_buffer,
            .meshlet_buffer = meshlet_buffer,
            .allocator = allocator,
        };
    }

    pub fn deinit(self: *MeshCache, vma_instance: vma.VmaAllocator) void {
        self.mesh2_instances.deinit(self.allocator);

        buffer_module.destroyBuffer(vma_instance, self.meshlet_buffer);
        buffer_module.destroyBuffer(vma_instance, self.vertex_attributes_buffer);
        buffer_module.destroyBuffer(vma_instance, self.vertex_buffer_position);
        buffer_module.destroyBuffer(vma_instance, self.index_buffer);
    }

    /// This invalidates all current handles.
    pub fn clearMeshes(self: *MeshCache) void {
        self.mesh2_instances.clearRetainingCapacity();

        self.current_index_offset = 0;
        self.current_position_offset = 0;
        self.current_attributes_offset = 0;
        self.current_meshlet_offset = 0;
    }

    fn allocateMesh(self: *MeshCache, mesh: Mesh, meshlets: []const Meshlet) MeshAlloc {
        const index_count: u32 = @intCast(mesh.indexes.items.len);
        const position_count: u32 = @intCast(mesh.positions.items.len);
        const attributes_count: u32 = @intCast(mesh.attributes.items.len);
        const meshlet_count: u32 = @intCast(meshlets.len);

        std.debug.assert(index_count > 0);
        std.debug.assert(position_count > 0);
        std.debug.assert(position_count == attributes_count);
        std.debug.assert(meshlet_count > 0);

        const alloc = MeshAlloc{
            .index_count = index_count,
            .vertex_count = position_count,
            .meshlet_count = meshlet_count,
            .index_offset = self.current_index_offset,
            .position_offset = self.current_position_offset,
            .attributes_offset = self.current_attributes_offset,
            .meshlet_offset = self.current_meshlet_offset,
        };

        self.current_index_offset += index_count;
        self.current_position_offset += position_count;
        self.current_attributes_offset += position_count;
        self.current_meshlet_offset += meshlet_count;

        std.debug.assert(self.current_index_offset < max_index_count);
        std.debug.assert(self.current_position_offset < max_vertex_count);
        std.debug.assert(self.current_attributes_offset < max_vertex_count);
        std.debug.assert(self.current_meshlet_offset < max_meshlet_count);

        return alloc;
    }
};

/// Mirrors load_meshes(): clusterize, allocate a slot, upload, hand back a
/// handle per input mesh.
pub fn loadMeshes(
    cache: *MeshCache,
    vma_instance: vma.VmaAllocator,
    allocator: std.mem.Allocator,
    meshes: []const Mesh,
    output_handles: []MeshHandle,
) !void {
    std.debug.assert(output_handles.len >= meshes.len);

    for (meshes, 0..) |mesh, mesh_index| {
        var meshlet_mesh = try meshlet_builder.buildMeshlets(allocator, mesh);
        defer meshlet_mesh.deinit(allocator);

        // allocateMesh reads counts off a Mesh, so wrap the flat slices back up.
        var view = Mesh{};
        view.indexes = .{ .items = meshlet_mesh.indexes, .capacity = meshlet_mesh.indexes.len };
        view.positions = .{ .items = meshlet_mesh.positions, .capacity = meshlet_mesh.positions.len };
        view.attributes = .{ .items = meshlet_mesh.attributes, .capacity = meshlet_mesh.attributes.len };

        const alloc = cache.allocateMesh(view, meshlet_mesh.meshlets);

        try uploadMeshToCache(cache, vma_instance, meshlet_mesh, alloc);

        const new_handle: MeshHandle = @enumFromInt(cache.mesh2_instances.items.len);
        try cache.mesh2_instances.append(cache.allocator, mesh2.createMesh2(alloc));

        log.debug("loaded mesh: {} indices, {} vertices, {} meshlets", .{
            alloc.index_count, alloc.vertex_count, alloc.meshlet_count,
        });

        output_handles[mesh_index] = new_handle;
    }
}

fn uploadMeshToCache(
    cache: *MeshCache,
    vma_instance: vma.VmaAllocator,
    meshlet_mesh: MeshletMesh,
    alloc: MeshAlloc,
) !void {
    try buffer_module.uploadBufferData(
        vma_instance,
        cache.index_buffer,
        cache.index_buffer.properties_deprecated,
        std.mem.sliceAsBytes(meshlet_mesh.indexes),
        alloc.index_offset,
    );

    try buffer_module.uploadBufferData(
        vma_instance,
        cache.vertex_buffer_position,
        cache.vertex_buffer_position.properties_deprecated,
        std.mem.sliceAsBytes(meshlet_mesh.positions),
        alloc.position_offset,
    );

    try buffer_module.uploadBufferData(
        vma_instance,
        cache.vertex_attributes_buffer,
        cache.vertex_attributes_buffer.properties_deprecated,
        std.mem.sliceAsBytes(meshlet_mesh.attributes),
        alloc.attributes_offset,
    );

    try buffer_module.uploadBufferData(
        vma_instance,
        cache.meshlet_buffer,
        cache.meshlet_buffer.properties_deprecated,
        std.mem.sliceAsBytes(meshlet_mesh.meshlets),
        alloc.meshlet_offset,
    );
}
